### ISSUE TITLE
Can't compile c10::optional values() or operator-> in .cu file

### ISSUE THREAD

**Issue by gchanan** [pytorch/pytorch#18496]
## 🐛 Bug

Our implementation of optional doesn't seem to always be compilable in .cu files.  Compiling this commit: https://github.com/gchanan/pytorch/commit/6666ff1083a55e90b230d2269c13a3d30af8c0f4

gives me the following error:
`
Optional.h(602): error: identifier "Typeinfo for  ::c10::bad_optional_access" is undefined in device code
`

I did a bit of digging and the problem seems to be with the constexpr functions, e.g.:
https://github.com/pytorch/pytorch/blob/654e59fcac4a9d4bf0b48306e1d7f7be5b7e40b1/c10/util/Optional.h#L600-L604

basically, for nvcc to compile this in a .cu file, `bad_optional_access(...)` has to be a `constexpr`.  This doesn't seem possible with the constraint that bad_optional_access derives from `std::exception` because `std::exception` does not have `constexpr` constructors.

Note that this only applies to calls to `value()`.  Calls to `operator->` seem to fail for other reasons that I didn't dig into.  But `operator*` passes fine, so the solution is to use that until we fix the issue.

cc @ngimel

**Comment by xsacha** [comment]
Is this on CUDA 10.1?

**Comment by ngimel** [comment]
Closing, 10.1 is no longer supported, and 10.2 is on the way out.
